File: src/test/java/persistence/sql/ddl/ColumnTest.java
package persistence.sql.ddl;

import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import static org.assertj.core.api.Assertions.assertThat;
import static org.junit.jupiter.api.Assertions.assertAll;
import static org.junit.jupiter.api.Assertions.assertThrows;

class ColumnTest {

    private static final Logger logger = LoggerFactory.getLogger(ColumnTest.class);

    @Test
    @DisplayName("Person Class로 Column 생성 : id Field에 대하여 name, Column Type, primary Key 여부가 잘 변환되었는지 확인")
    void createColumnClassWithPersonClass_id() throws Exception {
        ColumnInfo idColumn = ColumnInfo.extract(Person.class.getDeclaredField("id"));
        logger.info("{} 의 타입은 {}, primary 는 {}", idColumn.getName(), idColumn.getColumnType().getJavaType(), idColumn.isPrimaryKey());
        assertAll(
                () -> assertThat(idColumn.getColumnType()).isEqualTo(ColumnType.BIGINT),
                () -> assertThat(idColumn.getName()).isEqualTo("id"),
                () -> assertThat(idColumn.isPrimaryKey()).isTrue()
        );
    }

    @Test
    @DisplayName("Person Class로 Column 생성 : name Field에 대하여 name, Column Type, primary Key 여부가 잘 변환되었는지 확인")
    void createColumnClassWithPersonClass_name() throws Exception {
        ColumnInfo nameColumn = ColumnInfo.extract(Person.class.getDeclaredField("name"));
        logger.info("{} 의 타입은 {}, primary 는 {}", nameColumn.getName(), nameColumn.getColumnType().getJavaType(), nameColumn.isPrimaryKey());
        assertAll(
                () -> assertThat(nameColumn.getColumnType()).isEqualTo(ColumnType.VARCHAR),
                () -> assertThat(nameColumn.getName()).isEqualTo("nick_name"),
                () -> assertThat(nameColumn.isPrimaryKey()).isFalse()
        );
    }

    @Test
    @DisplayName("Person Class로 Column 생성 : age Field에 대하여 name, Column Type, primary Key 여부가 잘 변환되었는지 확인")
    void createColumnClassWithPersonClass_age() throws Exception {
        ColumnInfo ageColumn = ColumnInfo.extract(Person.class.getDeclaredField("age"));
        logger.info("{} 의 타입은 {}, primary 는 {}", ageColumn.getName(), ageColumn.getColumnType().getJavaType(), ageColumn.isPrimaryKey());
        assertAll(
                () -> assertThat(ageColumn.getColumnType()).isEqualTo(ColumnType.INTEGER),
                () -> assertThat(ageColumn.getName()).isEqualTo("old"),
                () -> assertThat(ageColumn.isPrimaryKey()).isFalse()
        );
    }
}

File: src/test/java/persistence/sql/ddl/ColumnTest.kt
package persistence.sql.ddl

import entity.Person
import exception.ColumnTypeUnavailableException
import io.kotest.assertions.throwables.shouldThrow
import io.kotest.core.spec.style.DescribeSpec
import io.kotest.matchers.shouldBe
import persistence.sql.QueryManagerFactory
import persistence.sql.ddl.column.Column
import persistence.sql.ddl.table.Table

class ColumnTest: DescribeSpec({

    describe("createQuery 메소드는") {
        val person = Person::class.java

        context("지원하는 Field 타입은") {

            val idColumn = QueryManagerFactory.column(person.getDeclaredField("id"))
            val stringColumn = QueryManagerFactory.column(person.getDeclaredField("name"))
            val intColumn = QueryManagerFactory.column(person.getDeclaredField("age"))
            it("DDL 컬럼 문자열을 반환한다.") {
                stringColumn.createQuery() shouldBe "nick_name VARCHAR(255) DEFAULT NULL"
                intColumn.createQuery() shouldBe "old int DEFAULT NULL"
                idColumn.createQuery() shouldBe "id bigint NOT NULL AUTO_INCREMENT, PRIMARY KEY (id)"
            }
        }

        context("지원하지 않는 Field 타입은") {
            val table = Table::class.java
            val unavailableColumn = QueryManagerFactory.column(table.getDeclaredField("tableParsers"))

            it("지원하지 않는 타입이라는 의미의 예외를 던진다.") {
                shouldThrow<ColumnTypeUnavailableException> { unavailableColumn.createQuery() }
            }
        }

        context("Transient 어노테이션이 작성된 필드는") {
            val transientColumn = QueryManagerFactory.column(person.getDeclaredField("index"))

            it("공백을 반환한다") {
                transientColumn.createQuery() shouldBe ""
            }
        }
    }
})
